File: mobile/flutter_app/lib/core/storage/storage_migration.dart
import 'package:shared_preferences/shared_preferences.dart';

/// The Questrace rename changed every stored key. Values written by earlier
/// builds are copied to the current keys on startup while the old keys are kept,
/// so rolling back to an earlier build still finds its session, draft and
/// preferences.
const legacyStoragePrefixes = <String, String>{
  'math-notebook:': 'questrace:',
  'notebook:': 'questrace:',
};

/// Returns the current key for [key], or null when the key is not renamed.
String? currentStorageKey(String key) {
  for (final entry in legacyStoragePrefixes.entries) {
    if (key.startsWith(entry.key)) {
      return '${entry.value}${key.substring(entry.key.length)}';
    }
  }
  return null;
}

/// Copies pre-rename entries to their current keys and returns the copied keys.
/// A value already stored under the current key always wins.
Future<List<String>> migrateLegacyStorage(SharedPreferences preferences) async {
  final pending = <String, Object>{};
  for (final key in preferences.getKeys()) {
    final current = currentStorageKey(key);
    if (current == null || preferences.containsKey(current)) {
      continue;
    }
    final value = preferences.get(key);
    if (value != null) {
      pending[current] = value;
    }
  }

  final copied = <String>[];
  for (final entry in pending.entries) {
    final value = entry.value;
    final stored = switch (value) {
      String() => await preferences.setString(entry.key, value),
      int() => await preferences.setInt(entry.key, value),
      bool() => await preferences.setBool(entry.key, value),
      double() => await preferences.setDouble(entry.key, value),
      List<String>() => await preferences.setStringList(entry.key, value),
      _ => false,
    };
    if (stored) {
      copied.add(entry.key);
    }
  }
  return copied;
}

File: frontend/src/utils/storage-migration.ts
// The Questrace rename changed every browser storage key. Values stored under
// the pre-rename keys are copied to the new keys on startup; the old keys are
// kept so that rolling back to an older build still finds its data.
const RENAMES: ReadonlyArray<readonly [prefix: string, replacement: string]> = [
  ['math-notebook:', 'questrace:'],
  ['notebook:', 'questrace:'],
]

export function currentKeyFor(key: string): string | null {
  const rename = RENAMES.find(([prefix]) => key.startsWith(prefix))
  return rename ? rename[1] + key.slice(rename[0].length) : null
}

/** Copies pre-rename entries of one storage area to their current keys. */
export function migrateLegacyKeys(storage: Storage): string[] {
  const pending: Array<[key: string, value: string]> = []
  for (let index = 0; index < storage.length; index += 1) {
    const key = storage.key(index)
    if (key === null) {
      continue
    }

    const current = currentKeyFor(key)
    if (current === null) {
      continue
    }

    // A value already written under the current key always wins.
    const value = storage.getItem(key)
    if (value !== null && storage.getItem(current) === null) {
      pending.push([current, value])
    }
  }

  for (const [key, value] of pending) {
    try {
      storage.setItem(key, value)
    } catch {
      // Private browsing can reject writes; the old key still holds the data.
    }
  }
  return pending.map(([key]) => key)
}

export function migrateLegacyStorage() {
  for (const area of [window.localStorage, window.sessionStorage]) {
    try {
      migrateLegacyKeys(area)
    } catch {
      // Storage can be unavailable entirely; the app then starts with defaults.
    }
  }
}

// Runs while this module is imported, before the appearance module reads the
// preferences it persisted.
migrateLegacyStorage()
